flang/lib/Lower/ConvertExprToHLFIR.cpp

```diff
@@ -1435,8 +1434,10 @@ struct UnaryOp<
       // allocate space on the stack for toBuffer
       auto dest = builder.create<fir::AllocaOp>(loc, toTy,
                                                 mlir::ValueRange{bufferSize});
-      builder.create<fir::CharConvertOp>(loc, lhs.getFirBase(), origBufferSize,
-                                         dest);
+      auto src = hlfir::convertToAddress(loc, builder, lhs,
+                                         lhs.getFortranElementType());
+      builder.create<fir::CharConvertOp>(loc, src.first.getCharBox()->getAddr(),
+                                         origBufferSize, dest);
```

Comment:
Please add a call to src.second here if any to insert hlfir.destroy/hlfir.end_associate if some temporary buffer was created for the source. Otherwise looks good.

Reply:
Added the call. What do you think?